### PR TITLE
Fix HTTP 401 `woocommerce_rest_cannot_view` `Sorry, you cannot list resources` REST API response

### LINKED ISSUE #26847: REST API Authentication does not work if wp_get_current_user() is called before WooCommerce is loaded
**Prerequisites (mark completed items with an [x]):**
- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
WooCommerce's REST API authentication relies on the `determine_current_user` function in order to perform its authentication ([ref](https://github.com/woocommerce/woocommerce/blob/4.2.1/includes/class-wc-rest-authentication.php#L41)). This filter runs whenever WordPress' `wp_get_current_user()` function is executed for the very first time during a request.

If a site has a plugin or code that calls `wp_get_current_user()` early during bootstrap (rather than waiting until `plugins_loaded`, then REST API requests return an HTTP 401 response:

```
{"code":"woocommerce_rest_cannot_view","message":"Sorry, you cannot list resources.","data":{"status":401}}
```

**Expected behavior**
I expect REST API requests to authenticate correctly.

**Actual behavior**
HTTP 401 reponses are received, even though a valid key and secret are provided via the URL parameters.

Our WooCommerce Zapier extension uses the WooCommerce REST API. We have approximately 2-3% of our user base unable to authenticate because their REST API returns HTTP 401 responses even though valid keys and secrets are provided.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
1. Create a REST API consumer key and secret via wp-admin --> WooCommerce --> Settings --> Advanced -->  REST API.
2. Load `https://example.com/wp-json/wc/v3/orders/?consumer_key=ck_***&consumer_secret=cs_***` in your browser or API client such as Insomnia.  Ensure the `consumer_key=` and `consumer_secret=` parameters in the URL are updated to use your newly created key and secret from step 1. Confirm that  HTTP 200 OK response is received, which demonstrates that HTTP basic authentication via query parameters is working as expected.
3. Create a `wp-content/mu-plugins/file.php` file with the following in it: 
```
require ABSPATH . WPINC . '/pluggable.php';
wp_get_current_user();
```
Alternatively, you could add the above code to any active plugin's main PHP file, provided that the plugin folder is alphabetically before `woocommerce`.
The above code causes `wp_get_current_user()` to be executed before WooCommerce is bootstrapped.
4. Refresh your browser/client from step 2. A HTTP 401 `woocommerce_rest_cannot_view` `Sorry, you cannot list resources` response is received.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
N/A
</details>

### PR DESCRIPTION
**Pull request by thejamescollins**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Fixes an edge-case that causes a `woocommerce_rest_cannot_view` `Sorry, you cannot list resources` HTTP 401 error response to occur during REST API requests on sites where a plugin or code determines the current WordPress user before WooCommerce is fully initialized.

Based on suggestions and feedback from @ObliviousHarmony (thanks Christopher!)

Closes #26847.

### How to test the changes in this Pull Request:

Please see #26847 for suggested steps to reproduce the bug and test that this change fixes the issue.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - Unexpected HTTP 401 `woocommerce_rest_cannot_view` `Sorry, you cannot list resources` REST API responses that occur when a plugin or custom code determines the current WordPress user before WooCommerce is fully initialized.

### CODE REVIEW

**Comment by om4james** [comment]
Hi @ObliviousHarmony I'd really appreciate it if you could review this PR as it is based on your ideas and feedback in #26847. 

Thank you 🙏

**Comment by ObliviousHarmony** [comment]
Hey @om4james,

Sorry for not better communicating my intentions here. This is on my short list to review it had just been a hectic week here in the Pacific Northwest!

**Comment by om4james** [comment]
No worries at all.

My main aim is to make sure it's on your radar, as I'd love it to be included in WC 5.6.

The issue is affecting some WooCommerce Zapier users, and quite likely a small portion of other REST API users. 

If you need anything from me please let me know.

Thanks

**Review by ObliviousHarmony — DISMISSED** [review]
Thanks for this @om4james, everything seems to work as expected. I've left a few thoughts but it looks great!

File: includes/class-wc-rest-authentication.php
Comment (om4james) [reply includes/class-wc-rest-authentication.php:42]: I think we should use the default priority (of 10) unless there is a technical reason to increase it to 15.

Specifying it as 15 just to match existing code is more of a vanity change that doesn't serve much of a purpose I think.

File: includes/class-wc-rest-authentication.php
Comment (om4james) [reply includes/class-wc-rest-authentication.php:104]: Are you suggesting that we add the following kind of check (similar to check_authentication_error()) at the beginning of this new authentication_fallback() method above the line 105 if statement?
```
		if ( ! empty( $error ) ) {
			return $error;
		}
```

**Review by thejamescollins — COMMENTED** [review]
Thanks for your feedback @ObliviousHarmony, I've responded to your comments.
James

File: includes/class-wc-rest-authentication.php
Comment (ObliviousHarmony) [reply includes/class-wc-rest-authentication.php:104]: If another plugin has already explicitly declared a failure we shouldn't waste our time trying to authenticate since the request has already failed.

Comment (om4james) [reply includes/class-wc-rest-authentication.php:104]: @ObliviousHarmony good thinking, I've made that change in ce3126045228c88c7f2fffd0785bba50d2dd6751

**Comment by om4james** [comment]
Hi @ObliviousHarmony ,

This PR is ready for someone else to review. Do you have any suggestions on who might be best to review and approve this so that it can be merged?

Note: The E2E tests are [failing in this PR](https://travis-ci.org/github/woocommerce/woocommerce/jobs/729239026), but they are also [failing in WC master](https://travis-ci.org/github/woocommerce/woocommerce/jobs/729757864). So the changes in this PR aren't causing any additional failures I don't think.

I'd really like to get this into WC 4.6 so that our affected WooCommerce Zapier users (and other REST API users) can benefit from this fix.

Thank you for your time.

James

**Comment by ObliviousHarmony** [comment]
Hello @om4james,

Unfortunately the PR freeze for 4.6 was last week and so this will have to wait until 4.7. I'll see about getting it a second review though so we can have that committed to.

**Comment by rrennick** [comment]
@om4james Do you have the steps to test this?

**Comment by om4james** [comment]
Hi @rrennick issue #26847 explains how to reproduce the bug.

Once you reproduce it against master using those steps, please check out this branch and verify that a 401 no longer occurs.

Thank you

**Review by vedanshujain — CHANGES_REQUESTED** [review]
Great work and testing well! Needs a small change, and will it be feasible to write some tests for this method.

File: includes/class-wc-rest-authentication.php
Comment (om4james) [reply includes/class-wc-rest-authentication.php:113]: Thanks @vedanshujain , I agree.
Based on https://developer.wordpress.org/reference/hooks/rest_authentication_errors/, I think the correct response is to return true, which indicates that the authentication method was used, and it succeeded.
I've gone ahead and implemented that in 79dd557b7708d0ae37b1ed63cfc3683f4d72c974.

**Comment by om4james** [comment]
> Great work and testing well! Needs a small change, and will it be feasible to write some tests for this method.

The challenge here is that the [WC_REST_Authentication class currently only has 3% coverage](https://codecov.io/gh/woocommerce/woocommerce/src/master/includes/class-wc-rest-authentication.php), even though the class has existed since WooCommerce 2.6.

The coverage on this class is incidental, there aren't any dedicated tests to test any of the functionality of the `WC_REST_Authentication` class.

I'm all for writing unit tests wherever practical, however in this case I feel as though it may be quite difficult and time consuming to do so.

Is it possible to perform the manual testing of this PR as I've outlined, and get it merged into the upcoming WC 4.7? 

The bug was reported 3+ months ago, and it is potentially affecting many WooCommerce REST API users. It's definitely affecting WooCommerce Zapier users who have reached out to us for support, and we don't currently have a solution or workaround for them until this is merged and released.

Thank you.

**Review by vedanshujain — APPROVED** [review]
LGTM!

**Comment by om4james** [comment]
Thank you @vedanshujain. And thank you @ObliviousHarmony for your guidance and feedback while we implemented a solution! 💯

**Pull request merged**
